File: go/examples/basic_usage.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/nodemavencom/proxy/go/nodemaven"
)

func main() {
	fmt.Println("NodeMaven Go SDK - Basic Usage Example")
	fmt.Println("=====================================")

	// Initialize client
	client, err := nodemaven.NewClient(&nodemaven.Config{
		// APIKey will be read from NODEMAVEN_APIKEY environment variable
	})
	if err != nil {
		log.Fatalf("Failed to initialize client: %v", err)
	}

	fmt.Println("✓ Client initialized successfully")

	ctx := context.Background()

	// Example 1: Basic proxy usage
	fmt.Println("\n1. Basic Proxy Usage")
	fmt.Println("-------------------")
	
	basicProxy, err := client.GetProxyConfig(nil)
	if err != nil {
		log.Printf("Failed to get basic proxy config: %v", err)
	} else {
		ip, err := testProxyConnection(basicProxy)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		} else {
			fmt.Printf("✓ Basic proxy working! Your IP: %s\n", ip)
		}
	}

	// Example 2: Geo-targeted proxy
	fmt.Println("\n2. Geo-targeted Proxy (US)")
	fmt.Println("-------------------------")
	
	usProxy, err := client.GetProxyConfig(&nodemaven.ProxyOptions{
		Country: "US",
	})
	if err != nil {
		log.Printf("Failed to get US proxy config: %v", err)
	} else {
		ip, err := testProxyConnection(usProxy)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		} else {
			fmt.Printf("✓ US proxy working! Your IP: %s\n", ip)
		}
	}

	// Example 3: Sticky session
	fmt.Println("\n3. Sticky Session")
	fmt.Println("----------------")
	
	sessionID := "example_session_" + nodemaven.GenerateSessionID()
	sessionProxy, err := client.GetProxyConfig(&nodemaven.ProxyOptions{
		Country: "US",
		Session: sessionID,
	})
	if err != nil {
		log.Printf("Failed to get session proxy config: %v", err)
	} else {
		// Make multiple requests with same session
		var ips []string
		for i := 1; i <= 3; i++ {
			ip, err := testProxyConnection(sessionProxy)
			if err != nil {
				fmt.Printf("Request %d failed: %v\n", i, err)
			} else {
				fmt.Printf("✓ Request %d successful! IP: %s\n", i, ip)
				ips = append(ips, ip)
			}

			// Small delay between requests
			if i < 3 {
				time.Sleep(200 * time.Millisecond)
			}
		}

		// Check if all IPs are the same (sticky session working)
		if len(ips) > 0 {
			allSame := true
			firstIP := ips[0]
			for _, ip := range ips {
				if ip != firstIP {
					allSame = false
					break
				}
			}

			if allSame {
				fmt.Printf("✓ Sticky session working! All requests used same IP: %s\n", firstIP)
			} else {
				fmt.Printf("Warning: Sticky session may not be working - got different IPs\n")
			}
		}
	}

	// Example 4: Get account information
	fmt.Println("\n4. Account Information")
	fmt.Println("---------------------")
	
	userInfo, err := client.GetUserInfo(ctx)
	if err != nil {
		log.Printf("Failed to get user info: %v", err)
	} else {
		fmt.Printf("✓ Email: %s\n", userInfo.Email)
		if userInfo.TrafficLimit > 0 {
			remaining := userInfo.TrafficLimit - userInfo.TrafficUsed
			if remaining < 0 {
				remaining = 0
			}
			fmt.Printf("✓ Data Remaining: %s\n", nodemaven.FormatBytes(remaining))
		}
	}

	fmt.Println("\nExample completed successfully!")
}

// testProxyConnection tests a proxy connection and returns the IP address
func testProxyConnection(proxyConfig *nodemaven.ProxyConfig) (string, error) {
	client := proxyConfig.HTTPClientWithTimeout(15 * time.Second)

	resp, err := client.Get("https://httpbin.org/ip")
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("failed to parse JSON: %w", err)
	}

	if origin, ok := result["origin"].(string); ok {
		return origin, nil
	}

	return "", fmt.Errorf("did not return IP in expected format")
}

File: javascript/examples/basic_usage.js
#!/usr/bin/env node

/**
 * NodeMaven JavaScript SDK - Basic Usage Example
 * 
 * This example demonstrates basic functionality of the NodeMaven JavaScript SDK:
 * - Client initialization
 * - Getting user information
 * - Retrieving location data (countries, regions, cities)
 * - Getting proxy configurations
 * - Error handling
 */

const { NodeMavenClient } = require('../src');

async function basicUsageExample() {
  try {
    console.log('🚀 NodeMaven JavaScript SDK - Basic Usage Example\n');

    // Initialize client (API key from environment variable NODEMAVEN_APIKEY)
    const client = new NodeMavenClient();
    console.log('✅ NodeMaven client initialized successfully');

    // Get user information
    console.log('\n📊 Getting user information...');
    const userInfo = await client.getUserInfo();
    console.log(`Username: ${userInfo.username}`);
    console.log(`Proxy Username: ${userInfo.proxy_username}`);
    console.log(`Traffic Used: ${userInfo.traffic_used} MB`);
    console.log(`Traffic Limit: ${userInfo.traffic_limit} MB`);

    // Get available countries (first 10)
    console.log('\n🌍 Getting available countries...');
    const countries = await client.getCountries({ limit: 10 });
    console.log(`Found ${countries.count} total countries. Showing first ${countries.results.length}:`);
    countries.results.forEach(country => {
      console.log(`  - ${country.name} (${country.code})`);
    });

    // Get regions for the first country
    if (countries.results.length > 0) {
      const firstCountry = countries.results[0];
      console.log(`\n🏘️  Getting regions for ${firstCountry.name}...`);
      const regions = await client.getRegions({ 
        countryCode: firstCountry.code, 
        limit: 5 
      });
      console.log(`Found ${regions.count} regions. Showing first ${regions.results.length}:`);
      regions.results.forEach(region => {
        console.log(`  - ${region.name} (${region.code})`);
      });

      // Get cities for the first region
      if (regions.results.length > 0) {
        const firstRegion = regions.results[0];
        console.log(`\n🏙️  Getting cities for ${firstRegion.name}...`);
        const cities = await client.getCities({
          countryCode: firstCountry.code,
          regionCode: firstRegion.code,
          limit: 5
        });
        console.log(`Found ${cities.count} cities. Showing first ${cities.results.length}:`);
        cities.results.forEach(city => {
          console.log(`  - ${city.name} (${city.code})`);
        });
      }
    }

    // Get proxy configuration
    console.log('\n🔧 Generating proxy configuration...');
    const proxyConfig = await client.getProxyConfig({
      country: 'US',
      city: 'new_york'
    });
    console.log('Proxy Configuration:');
    console.log(`  Host: ${proxyConfig.host}`);
    console.log(`  HTTP Port: ${proxyConfig.http_port}`);
    console.log(`  SOCKS5 Port: ${proxyConfig.socks5_port}`);
    console.log(`  Username: ${proxyConfig.username}`);
    console.log(`  Password: ${proxyConfig.password.substring(0, 8)}...`);

    // Get SOCKS5 proxy URL
    console.log('\n🔗 Generating SOCKS5 proxy URL...');
    const socks5Url = await client.getSocks5ProxyUrl({
      country: 'UK',
      city: 'london'
    });
    console.log(`SOCKS5 URL: ${socks5Url.substring(0, 50)}...`);

    // Get traffic statistics
    console.log('\n📈 Getting traffic statistics...');
    const statistics = await client.getStatistics({
      groupBy: 'day'
    });
    if (statistics.results && statistics.results.length > 0) {
      console.log(`Found ${statistics.results.length} statistics entries:`);
      statistics.results.slice(0, 3).forEach(stat => {
        console.log(`  - ${stat.date}: ${stat.traffic_used} MB`);
      });
    } else {
      console.log('No statistics data available');
    }

    console.log('\n✅ Basic usage example completed successfully!');

  } catch (error) {
    console.error('\n❌ Error occurred:');
    console.error(`Type: ${error.constructor.name}`);
    console.error(`Message: ${error.message}`);
    
    if (error.statusCode) {
      console.error(`Status Code: ${error.statusCode}`);
    }
    
    if (error.response) {
      console.error('Response Data:', JSON.stringify(error.response, null, 2));
    }
    
    process.exit(1);
  }
}

// Usage examples with different configurations
async function configurationExamples() {
  console.log('\n🔧 Configuration Examples:');
  
  // Example 1: Manual API key
  try {
    const client1 = new NodeMavenClient({
      apiKey: 'your-api-key-here'
    });
    console.log('✅ Client with manual API key');
  } catch (error) {
    console.log('⚠️  Manual API key example (expected to fail)');
  }
  
  // Example 2: Custom configuration
  const client2 = new NodeMavenClient({
    // apiKey: 'your-api-key',  // Can be set here or via environment
    baseUrl: 'https://dashboard.nodemaven.com',
    proxyHost: 'gate.nodemaven.com',
    httpPort: 8080,
    socks5Port: 1080,
    timeout: 30000
  });
  console.log('✅ Client with custom configuration');
}

async function main() {
  await basicUsageExample();
  await configurationExamples();
}

// Run the example if this file is executed directly
if (require.main === module) {
  main().catch(error => {
    console.error('❌ Example failed:', error.message);
    process.exit(1);
  });
}

module.exports = {
  basicUsageExample,
  configurationExamples
}; 